src/main/java/org/cbioportal/service/impl/ClinicalEventServiceImpl.java

```diff
@@ -108,4 +122,101 @@ public List<ClinicalEventTypeCount> getClinicalEventTypeCounts(List<String> stud
+        List<ClinicalEvent> endClinicalEventsMeta = survivalEvent.getEndClinicalEventsMeta();
+        ToIntFunction<ClinicalEvent> endPositionIdentifier = survivalEvent.getEndPositionIdentifier();
+        List<ClinicalEvent> censoredClinicalEventsMeta = survivalEvent.getCensoredClinicalEventsMeta();
+        ToIntFunction<ClinicalEvent> censoredPositionIdentifier = survivalEvent.getCensoredPositionIdentifier();
+
+        List<ClinicalEvent> patientStartEvents = clinicalEventRepository.getTimelineEvents(studyIds, patientIds, startClinicalEventsMeta);
+
+        List<String> filteredStudyIds = new ArrayList<>();
+        List<String> filteredPatientIds = new ArrayList<>();
+        for (ClinicalEvent clinicalEvent : patientStartEvents) {
+            filteredStudyIds.add(clinicalEvent.getStudyId());
+            filteredPatientIds.add(clinicalEvent.getPatientId());
+        }
+
+        // only fetch end timeline events for patients that have endClinicalEventsMeta and start timeline events
+        List<ClinicalEvent> patientEndEvents = new ArrayList<>();
+        if (CollectionUtils.isNotEmpty(endClinicalEventsMeta) && CollectionUtils.isNotEmpty(filteredStudyIds)) {
+            patientEndEvents = clinicalEventRepository.getTimelineEvents(filteredStudyIds, filteredPatientIds, endClinicalEventsMeta);
+        }
+
+        Map<String, ClinicalEvent> patientEndEventsById = patientEndEvents.stream().collect(Collectors.toMap(ClinicalEventServiceImpl::getKey, Function.identity()));
+
+        // filter out cases where start event is less than end event
+        patientStartEvents = patientStartEvents.stream()
+            .filter(event ->
+                Optional.ofNullable(patientEndEventsById.get(getKey(event)))
+                    .map(endPositionIdentifier::applyAsInt)
+                    .map(endDate -> startPositionIdentifier.applyAsInt(event) < endDate)
+                    .orElse(true)
+            ).toList();
+
+        filteredStudyIds.clear();
+        filteredPatientIds.clear();
+        // get all the studies that have start events but not end events
+        for (ClinicalEvent clinicalEvent : patientStartEvents.stream().filter(event -> !patientEndEventsById.containsKey(getKey(event))).toList()) {
+            filteredStudyIds.add(clinicalEvent.getStudyId());
+            filteredPatientIds.add(clinicalEvent.getPatientId());
+        }
+
+        List<ClinicalEvent> patientCensoredEvents = new ArrayList<>();
+        if (CollectionUtils.isNotEmpty(filteredStudyIds)) {
+            patientCensoredEvents = clinicalEventRepository.getTimelineEvents(filteredStudyIds, filteredPatientIds, censoredClinicalEventsMeta);
+        }
+
+        Map<String, ClinicalEvent> patientCensoredEventsById = patientCensoredEvents.stream().collect(Collectors.toMap(ClinicalEventServiceImpl::getKey, Function.identity()));
+
+        return patientStartEvents.stream()
```

Comment:
lambdas that have this many lines of code should be moved to a function

Reply:
extracted out into separate functions